app/src/main/java/com/tangem/tap/domain/card/DefaultDerivationsRepository.kt

```diff
@@ -0,0 +1,78 @@
+package com.tangem.tap.domain.card
+
+import arrow.core.Either
+import arrow.core.left
+import arrow.core.right
+import com.tangem.common.doOnFailure
+import com.tangem.common.doOnSuccess
+import com.tangem.common.extensions.ByteArrayKey
+import com.tangem.crypto.hdWallet.DerivationPath
+import com.tangem.datasource.local.userwallet.UserWalletsStore
+import com.tangem.domain.card.DerivePublicKeysError
+import com.tangem.domain.card.DerivePublicKeysError.*
+import com.tangem.domain.card.repository.DerivationsRepository
+import com.tangem.domain.tokens.model.CryptoCurrency
+import com.tangem.domain.wallets.models.UserWallet
+import com.tangem.domain.wallets.models.UserWalletId
+import com.tangem.operations.derivation.ExtendedPublicKeysMap
+import com.tangem.tap.domain.TangemSdkManager
+import com.tangem.utils.coroutines.CoroutineDispatcherProvider
+import com.tangem.utils.coroutines.runCatching
+
+internal typealias Derivations = Map<ByteArrayKey, List<DerivationPath>>
+private typealias DerivedKeys = Map<ByteArrayKey, ExtendedPublicKeysMap>
+
+internal class DefaultDerivationsRepository(
+    private val tangemSdkManager: TangemSdkManager,
+    private val userWalletsStore: UserWalletsStore,
+    private val dispatchers: CoroutineDispatcherProvider,
+) : DerivationsRepository {
+
+    override suspend fun derivePublicKeys(
+        userWalletId: UserWalletId,
+        currencies: List<CryptoCurrency>,
+    ): Either<DerivePublicKeysError, Unit> {
+        val userWallet = userWalletsStore.getSyncOrNull(userWalletId) ?: return UserWalletIdNotFound.left()
+        val derivations = MissedDerivationsFinder(scanResponse = userWallet.scanResponse).find(currencies)
+
+        if (derivations.isEmpty()) return NothingToDerive.left()
+
+        tangemSdkManager.derivePublicKeys(cardId = null, derivations = derivations)
+            .doOnFailure { return UnableToDerive.left() }
+            .doOnSuccess {
+                updatePublicKeys(userWalletId = userWalletId, keys = it.entries).fold(
+                    onSuccess = { return Unit.right() },
+                    onFailure = { return UnableToDerive.left() },
+                )
+            }
+
+        error("This code should never be reached")
+    }
```

Comment:
Предлагаю упростить:
```suggestion
    override suspend fun derivePublicKeys(
        userWalletId: UserWalletId,
        currencies: List<CryptoCurrency>
    ) {
        val userWallet = userWalletsStore.getSyncOrNull(userWalletId) ?: error("User wallet not found")
        val derivations = MissedDerivationsFinder(scanResponse = userWallet.scanResponse).find(currencies)
        
        if (derivations.isEmpty()) return

        val derivationResult = tangemSdkManager
            .derivePublicKeys(cardId = null, derivations)
            .let {
                when (it) {
                    is CompletionResult.Failure -> throw it.error
                    is CompletionResult.Success -> it.data
                }
            }

        userWalletsStore.update(
            userWalletId = userWalletId,
            update = { it.updateDerivedKeys(derivationResult.entries) },
        )
    }
```

Reply:
А типизация ошибок?

Reply:
Тут их 2 только:
1. Нечего дерайвить - нам не нужно её в логикие как-то обрабатывать, кажется, что достаточно просто вернуть успех
2. Эта маппится просто во `Throwable` по нему легче понять, что пошло не так, чем от простого объекта `UnableToDerive`